jetty-websocket/websocket-javax-tests/src/test/java/org/eclipse/jetty/websocket/javax/tests/SingleMessageHandlerTest.java

```diff
@@ -0,0 +1,139 @@
+import javax.websocket.OnMessage;
+import javax.websocket.Session;
+import javax.websocket.server.ServerEndpoint;
+
+import org.eclipse.jetty.server.Server;
+import org.eclipse.jetty.server.ServerConnector;
+import org.eclipse.jetty.servlet.ServletContextHandler;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.websocket.javax.client.internal.JavaxWebSocketClientContainer;
+import org.eclipse.jetty.websocket.javax.server.config.JavaxWebSocketServletContainerInitializer;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class SingleMessageHandlerTest
+{
+    private static final LinkedBlockingQueue<ByteBuffer> BINARY_MESSAGES = new LinkedBlockingQueue<>();
+    private static final LinkedBlockingQueue<String> TEXT_MESSAGES = new LinkedBlockingQueue<>();
+
+    private Server _server;
+    private ServerConnector _connector;
+    private JavaxWebSocketClientContainer _client;
+
+    @BeforeEach
+    public void before() throws Exception
+    {
+        _server = new Server();
+        _connector = new ServerConnector(_server);
+        _server.addConnector(_connector);
+
+        ServletContextHandler contextHandler = new ServletContextHandler();
+        JavaxWebSocketServletContainerInitializer.configure(contextHandler, ((servletContext, serverContainer) ->
+        {
+            serverContainer.addEndpoint(TextEndpoint.class);
+            serverContainer.addEndpoint(BinaryEndpoint.class);
+        }));
+        _server.setHandler(contextHandler);
+
+        _server.start();
+        _client = new JavaxWebSocketClientContainer();
+        _client.start();
+    }
+
+    @ServerEndpoint("/binary")
+    public static class BinaryEndpoint
+    {
+        @OnMessage
+        public void onMessage(ByteBuffer message)
+        {
+            BINARY_MESSAGES.add(message);
+        }
+    }
+
+    @ServerEndpoint("/text")
+    public static class TextEndpoint
+    {
+        @OnMessage
+        public void onMessage(String message)
+        {
+            TEXT_MESSAGES.add(message);
+        }
+    }
+
+    @AfterEach
+    public void after() throws Exception
+    {
+        _client.stop();
+        _server.stop();
+    }
+
+    @Test
+    public void testBinary() throws Exception
+    {
+        URI uri = URI.create("ws://localhost:" + _connector.getLocalPort() + "/binary");
+        EventSocket eventSocket = new EventSocket();
+        Session session = _client.connectToServer(eventSocket, uri);
+
+        // Can send/receive binary message successfully.
+        ByteBuffer binaryMessage = BufferUtil.toBuffer("hello world");
+        session.getBasicRemote().sendBinary(binaryMessage);
+        assertThat(BINARY_MESSAGES.poll(5, TimeUnit.SECONDS), is(binaryMessage));
```

Comment:
This is so confusing. That `is(binaryMessage)` actually calls `equals()` is so weird. Why not using `assertEquals()`?

Also, why `binaryMessage` is not consumed after `sendBinary()`?!? This seems like a bug, I wonder how the test is passing?

Reply:
> This is so confusing. That is(binaryMessage) actually calls equals() is so weird. Why not using assertEquals()?

I don't see this as confusing, this pattern is frequently used in tests.

> Also, why binaryMessage is not consumed after sendBinary()?!? This seems like a bug, I wonder how the test is passing?

This was an intentional change that was made. See issue #4341. @gregw has previously said that sending a frame should not modify the position/limit of the original buffer.